File: typescript/tests/client.test.ts
/**
 * Tests for OpenContextClient – mocked HTTP layer so no server needed.
 */

import { OpenContextClient } from "../src/client";

// Mock global fetch
const mockFetch = jest.fn();
global.fetch = mockFetch as typeof fetch;

function mockOk(body: unknown): void {
  mockFetch.mockResolvedValueOnce({
    ok: true,
    json: () => Promise.resolve(body),
    text: () => Promise.resolve(JSON.stringify(body)),
  } as unknown as Response);
}

function mockError(status: number, text: string): void {
  mockFetch.mockResolvedValueOnce({
    ok: false,
    status,
    json: () => Promise.resolve({ detail: text }),
    text: () => Promise.resolve(text),
  } as unknown as Response);
}

beforeEach(() => {
  mockFetch.mockClear();
});

describe("OpenContextClient", () => {
  const client = new OpenContextClient({ userId: "test-user" });

  // ---------------------------------------------------------------- //
  // augment
  // ---------------------------------------------------------------- //

  test("augment sends correct payload", async () => {
    const augmented = [
      { role: "system" as const, content: "## Context\n- Fact: I use Python" },
      { role: "user" as const, content: "Hello" },
    ];
    mockOk({ messages: augmented });

    const result = await client.augment({
      messages: [{ role: "user", content: "Hello" }],
      sessionId: "sess-1",
    });

    expect(result.messages).toHaveLength(2);
    expect(mockFetch).toHaveBeenCalledTimes(1);
    const [url, opts] = mockFetch.mock.calls[0] as [string, RequestInit];
    expect(url).toBe("http://localhost:8765/api/v1/augment");
    const body = JSON.parse(opts.body as string);
    expect(body.user_id).toBe("test-user");
    expect(body.session_id).toBe("sess-1");
  });

  // ---------------------------------------------------------------- //
  // record
  // ---------------------------------------------------------------- //

  test("record sends correct payload", async () => {
    mockOk({ session_id: "sess-1", message_count: 2 });

    const result = await client.record({
      sessionId: "sess-1",
      messages: [
        { role: "user", content: "I love TypeScript" },
        { role: "assistant", content: "Great!" },
      ],
    });

    expect(result.session_id).toBe("sess-1");
    expect(result.message_count).toBe(2);
    const [, opts] = mockFetch.mock.calls[0] as [string, RequestInit];
    const body = JSON.parse(opts.body as string);
    expect(body.session_id).toBe("sess-1");
    expect(body.messages).toHaveLength(2);
  });

  // ---------------------------------------------------------------- //
  // getContext
  // ---------------------------------------------------------------- //

  test("getContext builds correct query string", async () => {
    mockOk({ injected_system_prompt: "## Context", memories: [], relevance_scores: {} });

    await client.getContext({ query: "Python tips" });

    const [url] = mockFetch.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/context");
    expect(url).toContain("query=Python+tips");
    expect(url).toContain("user_id=test-user");
  });

  // ---------------------------------------------------------------- //
  // addMemory / getMemories / deleteMemory
  // ---------------------------------------------------------------- //

  test("addMemory sends correct payload", async () => {
    const mem = {
      id: "mem-1",
      content: "I know Rust",
      memory_type: "skill",
      importance: 0.8,
      tags: ["rust"],
      source_session_id: null,
      created_at: "2024-01-01T00:00:00",
      updated_at: "2024-01-01T00:00:00",
      last_accessed: "2024-01-01T00:00:00",
      access_count: 0,
      metadata: {},
    };
    mockOk(mem);

    const result = await client.addMemory({ content: "I know Rust", memoryType: "skill" });

    expect(result.id).toBe("mem-1");
    const [, opts] = mockFetch.mock.calls[0] as [string, RequestInit];
    const body = JSON.parse(opts.body as string);
    expect(body.content).toBe("I know Rust");
    expect(body.memory_type).toBe("skill");
  });

  test("getMemories includes memory_type filter when provided", async () => {
    mockOk({ memories: [], count: 0 });

    await client.getMemories({ memoryType: "skill" });

    const [url] = mockFetch.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("memory_type=skill");
  });

  test("deleteMemory calls DELETE endpoint", async () => {
    mockOk({ deleted: true, memory_id: "mem-abc" });

    const result = await client.deleteMemory("mem-abc");

    expect(result.deleted).toBe(true);
    const [url, opts] = mockFetch.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/memories/mem-abc");
    expect(opts.method).toBe("DELETE");
  });

  // ---------------------------------------------------------------- //
  // getUserProfile / getSessions / clearUserData
  // ---------------------------------------------------------------- //

  test("getUserProfile calls correct endpoint", async () => {
    mockOk({ user_id: "test-user", name: "Alice", skills: [], preferences: {}, habits: [], facts: [], updated_at: "2024-01-01T00:00:00", exists: true });

    const profile = await client.getUserProfile();
    expect(profile.user_id).toBe("test-user");
    const [url] = mockFetch.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/profile");
  });

  test("clearUserData calls DELETE on user endpoint", async () => {
    mockOk({ cleared: true, user_id: "test-user" });

    const result = await client.clearUserData();
    expect(result.cleared).toBe(true);
    const [url, opts] = mockFetch.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/users/test-user");
    expect(opts.method).toBe("DELETE");
  });

  // ---------------------------------------------------------------- //
  // Error handling
  // ---------------------------------------------------------------- //

  test("throws on non-ok response", async () => {
    mockError(404, "Memory not found");

    await expect(client.deleteMemory("ghost-id")).rejects.toThrow(
      "openContext API error 404"
    );
  });

  test("throws when userId is missing", () => {
    const noUserClient = new OpenContextClient();
    expect(() => noUserClient["_userId"](undefined)).toThrow("userId is required");
  });
});

File: golang/opencontext/client_test.go
package opencontext_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	opencontext "github.com/jjh886/openContext/golang/opencontext"
)

// ------------------------------------------------------------------ //
// Helpers
// ------------------------------------------------------------------ //

func newTestServer(handler http.HandlerFunc) (*httptest.Server, *opencontext.Client) {
	srv := httptest.NewServer(handler)
	client := opencontext.NewClient(opencontext.Options{
		UserID:    "test-user",
		ServerURL: srv.URL,
	})
	return srv, client
}

func writeJSON(w http.ResponseWriter, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(v)
}

// ------------------------------------------------------------------ //
// Tests
// ------------------------------------------------------------------ //

func TestAugment(t *testing.T) {
	srv, client := newTestServer(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/api/v1/augment" {
			http.Error(w, "unexpected", http.StatusBadRequest)
			return
		}
		var req struct {
			UserID   string              `json:"user_id"`
			Messages []opencontext.Message `json:"messages"`
		}
		_ = json.NewDecoder(r.Body).Decode(&req)
		if req.UserID != "test-user" {
			http.Error(w, "wrong user_id", http.StatusBadRequest)
			return
		}
		writeJSON(w, map[string]interface{}{
			"messages": []map[string]string{
				{"role": "system", "content": "## Context"},
				{"role": "user", "content": "Hello"},
			},
		})
	})
	defer srv.Close()

	result, err := client.Augment(context.Background(), opencontext.AugmentRequest{
		Messages: []opencontext.Message{{Role: "user", Content: "Hello"}},
	})
	if err != nil {
		t.Fatalf("Augment error: %v", err)
	}
	if len(result.Messages) != 2 {
		t.Errorf("expected 2 messages, got %d", len(result.Messages))
	}
	if result.Messages[0].Role != "system" {
		t.Errorf("expected first message to be system, got %q", result.Messages[0].Role)
	}
}

func TestRecord(t *testing.T) {
	srv, client := newTestServer(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/api/v1/record" {
			http.Error(w, "unexpected", http.StatusBadRequest)
			return
		}
		writeJSON(w, map[string]interface{}{
			"session_id":    "sess-1",
			"message_count": 2,
		})
	})
	defer srv.Close()

	result, err := client.Record(context.Background(), opencontext.RecordRequest{
		SessionID: "sess-1",
		Messages: []opencontext.Message{
			{Role: "user", Content: "I use Go"},
			{Role: "assistant", Content: "Great!"},
		},
	})
	if err != nil {
		t.Fatalf("Record error: %v", err)
	}
	if result.SessionID != "sess-1" {
		t.Errorf("expected session-id sess-1, got %q", result.SessionID)
	}
	if result.MessageCount != 2 {
		t.Errorf("expected message_count 2, got %d", result.MessageCount)
	}
}

func TestAddMemory(t *testing.T) {
	srv, client := newTestServer(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/api/v1/memories" {
			http.Error(w, "unexpected", http.StatusBadRequest)
			return
		}
		var req struct {
			UserID  string `json:"user_id"`
			Content string `json:"content"`
		}
		_ = json.NewDecoder(r.Body).Decode(&req)
		writeJSON(w, map[string]interface{}{
			"id":               "mem-1",
			"content":          req.Content,
			"memory_type":      "fact",
			"importance":       0.7,
			"tags":             []string{},
			"source_session_id": nil,
			"created_at":       "2024-01-01T00:00:00",
			"updated_at":       "2024-01-01T00:00:00",
			"last_accessed":    "2024-01-01T00:00:00",
			"access_count":     0,
			"metadata":         map[string]interface{}{},
		})
	})
	defer srv.Close()

	mem, err := client.AddMemory(context.Background(), opencontext.AddMemoryRequest{
		Content: "I live in Berlin",
	})
	if err != nil {
		t.Fatalf("AddMemory error: %v", err)
	}
	if mem.ID != "mem-1" {
		t.Errorf("expected memory id mem-1, got %q", mem.ID)
	}
}

func TestGetMemories(t *testing.T) {
	srv, client := newTestServer(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/api/v1/memories" {
			http.Error(w, "unexpected", http.StatusBadRequest)
			return
		}
		writeJSON(w, map[string]interface{}{
			"memories": []interface{}{},
			"count":    0,
		})
	})
	defer srv.Close()

	result, err := client.GetMemories(context.Background(), "", 0)
	if err != nil {
		t.Fatalf("GetMemories error: %v", err)
	}
	if result.Count != 0 {
		t.Errorf("expected count 0, got %d", result.Count)
	}
}

func TestClearUserData(t *testing.T) {
	srv, client := newTestServer(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete {
			http.Error(w, "expected DELETE", http.StatusBadRequest)
			return
		}
		writeJSON(w, map[string]interface{}{"cleared": true, "user_id": "test-user"})
	})
	defer srv.Close()

	if err := client.ClearUserData(context.Background()); err != nil {
		t.Fatalf("ClearUserData error: %v", err)
	}
}

func TestServerError(t *testing.T) {
	srv, client := newTestServer(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "not found", http.StatusNotFound)
	})
	defer srv.Close()

	_, err := client.GetUserProfile(context.Background())
	if err == nil {
		t.Fatal("expected error, got nil")
	}
}
